### ISSUE TITLE
Separate response duplicates mid

### ISSUE THREAD

**Issue by obgm** [obgm/libcoap#726]
## Environment

- libcoap version v4.3.0-12-g1385ded9
- Build System:             both
- Operating System:         all
- Hosted Environment:      None

## Problem Description

`coap_register_async()` deliberately copies the original request's message id (mid) for response generation (see  [line 41](https://github.com/obgm/libcoap/blob/a80d462ff57630ce214efdf5caf34133b02ad7ee/src/async.c#L41) and [line 75](https://github.com/obgm/libcoap/blob/a80d462ff57630ce214efdf5caf34133b02ad7ee/src/async.c#L75), respectively).

Now, when `handle_request()` generates an empty ACK ([net.c, line 2520](https://github.com/obgm/libcoap/blob/a80d462ff57630ce214efdf5caf34133b02ad7ee/src/net.c#L2520)), any following responses must have a distinct message id as they form a new message transaction. `coap_check_async()` passes the copied pdu with the original mid when calling `handle_request()`, see [net.c line 3206](https://github.com/obgm/libcoap/blob/a80d462ff57630ce214efdf5caf34133b02ad7ee/src/net.c#L3206).

### Expected Behavior

Being separate transactions on the message layer, the separate response must have a distinct message id, cf. [Section 5.2.2 of RFC 7252](https://tools.ietf.org/rfcmarkup?doc=7252#section-5.2.2). Figure 20 in [Appendix A of RFC 7252](https://tools.ietf.org/rfcmarkup?doc=7252#appendix-A) illustrates this:

```
  Client  Server
      |      |
      |      |
      +----->|     Header: GET (T=CON, Code=0.01, MID=0x7d38)
      | GET  |      Token: 0x53
      |      |   Uri-Path: "temperature"
      |      |
      |      |
      |<- - -+     Header: (T=ACK, Code=0.00, MID=0x7d38)
      |      |
      |      |
      |<-----+     Header: 2.05 Content (T=CON, Code=2.05, MID=0xad7b)
      | 2.05 |      Token: 0x53
      |      |    Payload: "22.3 C"
      |      |
      |      |
      +- - ->|     Header: (T=ACK, Code=0.00, MID=0xad7b)
      |      |

             Figure 20: Confirmable Request; Separate Response
```

### Actual Behavior

As shown below, the separate response also has `i:4966` of the original request:

```
$ coap-client -v9 coap://libcoap.net/async
Jul 21 09:46:38.640 DEBG added 6 protocols, 52 ciphersuites, 19 sig algos and 10 groups into priority list
Jul 21 09:46:38.640 DEBG ***[source]:33268 <-> [libcoap.net]:5683 UDP : new outgoing session
Jul 21 09:46:38.640 DEBG sending CoAP request:
Jul 21 09:46:38.640 DEBG *  [source]:33268 <-> [libcoap.net]:5683 UDP : sent 22 bytes
v:1 t:CON c:GET i:4966 {} [ Uri-Host:libcoap.net, Uri-Path:async ]
Jul 21 09:46:38.640 DEBG ** [source]:33268 <-> [libcoap.net]:5683 UDP : mid=0x4966: added to retransmit queue (2875ms)
Jul 21 09:46:38.640 DEBG timeout is set to 90 seconds
Jul 21 09:46:38.674 DEBG *  [source]:33268 <-> [libcoap.net]:5683 UDP : received 4 bytes
v:1 t:ACK c:0.00 i:4966 {} [ ]
Jul 21 09:46:38.674 DEBG ** [source]:33268 <-> [libcoap.net]:5683 UDP : mid=0x4966: removed
Jul 21 09:46:43.673 DEBG *  [source]:33268 <-> [libcoap.net]:5683 UDP : received 10 bytes
v:1 t:ACK c:2.05 i:4966 {} [ Content-Format:text/plain ] :: 'done'
Jul 21 09:46:43.674 DEBG ** process incoming 2.05 response:
doneJul 21 09:46:43.674 DEBG ***[source]:33268 <-> [libcoap.net]:5683 UDP : session closed
```

### Steps to reproduce

1. Start `coap-server`, and ideally wireshark to capture on port 5683.
2. Run `coap-client -v9 coap://localhost/async`.

**Comment by mrdeep1** [comment]
@obgm Looks like my mistake - do you want me to create a fix for it?

**Comment by obgm** [comment]
Not really a mistake, just special case treatment missing. I have overlooked it in my review as well.
If you have an idea how to fix this properly, I would appreciate that. Thinking of the implications, I have not yet come to a conclusion on the best way to fix this:
1. Create a new mid for the response when an empty ACK was created (retransmissions of the original request would require any ACK to have the original mid), or
2. Store the original mid and re-patch it into the response whenever an empty ACK or a piggy-backed response is created?

**Comment by mrdeep1** [comment]
I'm working on something else at present - will try to look at it/think it through this afternoon, certainly tomorrow afternoon.

**Comment by obgm** [comment]
Thanks. I am on something else right now as well, so no hurry.

**Comment by mrdeep1** [comment]
See PR #727.

**Comment by obgm** [comment]
Closed after PR #727 has been merged.
